bpf/cpu/cpu.bpf.c

```diff
@@ -675,19 +679,82 @@ int walk_user_stacktrace_impl(struct bpf_perf_event_data *ctx) {
   }
 
   for (int i = 0; i < MAX_STACK_DEPTH_PER_PROGRAM; i++) {
+    LOG("[debug] Within unwinding machinery loop");
     LOG("## frame: %d", unwind_state->stack.len);
 
     LOG("\tcurrent pc: %llx", unwind_state->ip);
     LOG("\tcurrent sp: %llx", unwind_state->sp);
     LOG("\tcurrent bp: %llx", unwind_state->bp);
 
     u64 offset = 0;
+
     chunk_info_t *chunk_info = NULL;
     enum find_unwind_table_return unwind_table_result = find_unwind_table(&chunk_info, user_pid, unwind_state->ip, &offset);
 
+
     if (unwind_table_result == FIND_UNWIND_JITTED) {
-      LOG("JIT section, stopping");
-      return 1;
+      if (!unwinder_config.mixed_stack_walking) {
+        LOG("JIT section, stopping");
+        return 1;
+      }
+
+      LOG("[debug] Unwinding JITed stacks");
+
+      unwind_state->jitted = true;
+
+      u64 next_fp = 0;
+      u64 ra = 0;
+      u64 len = unwind_state->stack.len;
+
+      // When we enter a JITed stack, the first JITed frame can
+      // be obtained from the current value of pc(program counter)
+
+      if (unwind_state->stack.len == 0) {
+        if (len >= 0 && len < MAX_STACK_DEPTH) {
+          unwind_state->stack.addresses[len] = unwind_state->ip;
+          unwind_state->stack.len++;
+          continue;
+        }
+      }
+
+      err = bpf_probe_read_user(&next_fp, 8, (void *)unwind_state->bp);
+      if (err < 0) {
+        //TODO(sylfrena):
+        //for some weird reason commenting out this and the next err log line results in a panic
+        LOG("[debug] i=%d, err = %d && rbp = %llx && ra=%llx", i, err, next_fp, ra);
+        return false;
+      }
+
+      //LOG("[debug]  i=%d, err = %d && rbp = %llx && ra=%llx", i, err, next_fp, ra);
+
+      // reading return address
+      err = bpf_probe_read_user(&ra, 8, (void *)unwind_state->bp + 8);
+      if (err < 0) {
+        //TODO(sylfrena)
+        //for some weird reason commenting out this and the next err log line results in a panic
+        LOG("[debug] reading ra i=%d, err = %d && rbp = %llx && ra=%llx", i, err, next_fp, ra);
+        return false;
+      }
+
+      if (next_fp == 0) {
+        LOG("[info] found bottom frame while walking JITed section");
+        return true;
+      }
+
+      // TODO(sylfrena): add comments to explain calculations
+      unwind_state->sp = unwind_state->bp + 16;
+      unwind_state->bp = next_fp;
+      unwind_state->ip = ra - 1;
+      len = unwind_state->stack.len;
+
+      // add ra for frame
+      if (len >= 0 && len < MAX_STACK_DEPTH) {
+        unwind_state->stack.addresses[len] = ra;
+      }
+
+      unwind_state->stack.len++;
```

Comment:
nit: Can you move this inside the bound checks? Sorry I left it outside before you touched the code!